src/lem/circuit.rs

```diff
@@ -0,0 +1,902 @@
+//! related to each visited slot.
+//!
+//! * STEP 3: During construction of constraints, we first preallocate the preimage
+//! and image for each slot; then we synthesize all slots; and finally we traverse
+//! LEM (for the third time), adding implication constraints, glueing together all the
+//! information. This step is better explained in the `synthesize` function.
+
+use std::collections::{HashMap, HashSet};
+
+use anyhow::{anyhow, bail, Context, Result};
+use bellperson::{
+    gadgets::{boolean::Boolean, num::AllocatedNum},
+    ConstraintSystem,
+};
+use indexmap::{IndexMap, IndexSet};
+
+use crate::circuit::gadgets::{
+    constraints::{
+        alloc_equal_const, and, enforce_selector_with_premise, implies_equal, implies_equal_zero,
+    },
+    data::hash_poseidon,
+    pointer::AllocatedPtr,
+};
+
+use crate::field::{FWrap, LurkField};
+
+use super::{
+    interpreter::{Frame, SlotType},
+    path::Path,
+    pointers::ZPtr,
+    store::Store,
+    MetaPtr, LEM, LEMOP,
+};
+
+/// Holds a counter per path
+#[derive(Default)]
+struct PathTicker(HashMap<Path, usize>);
+
+impl PathTicker {
+    /// Increments the counter of a path. If the path wasn't tracked, returns
+    /// `0` and starts tracking it
+    pub(crate) fn tick(&mut self, path: Path) -> usize {
+        let next = self.0.get(&path).unwrap_or(&0).to_owned();
+        self.0.insert(path, next + 1);
+        next
+    }
+
+    /// Starts tracking a new path with a counter from another. If the reference
+    /// path wasn't being tracked, the new one won't be either, such that calling
+    /// `next` will return `0`.
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
+        match self.0.get(from) {
+            Some(i) => {
+                self.0.insert(new, *i);
+            }
+            None => (),
+        }
+    }
+}
+
+/// Keeps track of previously used slots indices for each possible LEM path,
+/// being capable of informing the next slot index to be used
+#[derive(Default)]
+struct SlotsCounter {
+    hash2: PathTicker,
+    hash3: PathTicker,
+    hash4: PathTicker,
+}
+
+impl SlotsCounter {
+    #[inline]
+    pub(crate) fn next_hash2(&mut self, path: Path) -> usize {
+        self.hash2.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash3(&mut self, path: Path) -> usize {
+        self.hash3.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash4(&mut self, path: Path) -> usize {
+        self.hash4.tick(path)
+    }
+
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
+        self.hash2.cont(new.clone(), from);
+        self.hash3.cont(new.clone(), from);
+        self.hash4.cont(new, from);
+    }
+}
+
+/// Contains a `slots_map` that maps `LEMOP`s to their slots. This map is not
+/// expected to be injective, as two or more `LEMOP`s can be mapped to the same
+/// slot.
+///
+/// The `slots` attribute is derived from `slots_map` by iterating on its values.
```

Comment:
```
match_tag qqq
  foo => let x = hash2(a, b)
  bar => let y = hash2(x, y)
```
We want those two operations to be mapped to the same slot. Only one call to the `hash_poseidon` gadget should happen.

I'm not sure if that answered your question though

Reply:
I think I may have gotten it?

 in my example you’re going to reserve 3 slots (because path1 needs as many), but will only compute something for those three slots in the case where path1 is actually concrete. If the concrete path is path[2-5] then the last of the three slots will be a dummy and the first two will be actually computed.
 
 If that’s correct, then it’s perhaps worth running through that example in a comment ?